### PR TITLE
Add support for drawing histogram with pre-aggregated (buckets + counts) data

### PR DESCRIPTION
**Pull request by Kami**
### Background, Context

A lot of Python libraries such as OpenTelemetry Metrics API (https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md) support histograms.

The problem is that those libraries expose pre-aggregated data (counts for each bucket) and not raw data. This is done intentionally since storing raw data could be very expensive in terms of memory usage in case there are a lot of samples.

Here is an example of data available by OpenTelemetry metrics Histogram view:

```python
 {
  'attributes': {},
   'bucket_counts': [1945, 0, 0, 0, 0, 0, 10555, 798, 0, 28351, 0],
   'count': 41649,
   'explicit_bounds': [10, 50, 100, 200, 300, 500, 800, 1000, 2000, 10000],
 }
```

As you can see, we get we get counts for each bucket. Buckets go from (n[i], n[i+1]) whereas first item is -infinity and the last item is +infinity. In this example, buckets would look something like this: ``[(-inf, 10), (11, 20), (21, 50), ..., (10000, +inf)]``.

## Proposed Change / Implementation

In this change I added a new function for graphing data which is already split into counts and bins.

This pull request is not final (I just copied the existing code + made small change to show how it should work), I just wanted to get the discussion going (with some code and examples) so we can agree on the API which should be exposed.

With the sample data above, this new function and bins which look like this:

```python
counts = bucket_counts
bins = [float("+inf")]
bins.extend(explicit_bounds)
bins.append(float("+inf"))
```

We get the following histogram:

```bash
        bucket       | ________________________________________________________________________________ Total Counts
[inf     ,       10) | ⣿⣿⣿⣿⣿⣿⣿⣿⣿⠆⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀ 3269
[10      ,       50) | ⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀ 0
[50      ,      100) | ⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀ 0
[100     ,      200) | ⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀ 0
[200     ,      300) | ⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀ 0
[300     ,      500) | ⣿⣿⣿⣿⣿⠆⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀ 1890
[500     ,      800) | ⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⡇⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀ 9706
[800     ,     1000) | ⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀ 0
[1000    ,     2000) | ⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⡗⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀ 8709
[2000    ,    10000) | ⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⣿⠀ 28213
[10000   ,      inf) | ⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀ 0
‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾
```

Feedback is welcome, especially on naming and things like that.

## TODO

- [ ] Agree on the API / implementation
- [ ] Finish implementing changes
- [ ] Tests
- [ ] Documentation
- [ ] Examples

### CODE REVIEW

File: plotille/_graphs.py [hunk 1]
```diff
@@ -31,6 +31,52 @@
 from ._input_formatter import InputFormatter
 from ._util import hist as compute_hist
 
+def hist_pre_aggregated(counts, bins, width=80, log_scale=False, linesep=os.linesep,
```
Comment (tammoippen) [inline plotille/_graphs.py:34]:
```suggestion
def hist_aggregated(counts, bins, width=80, log_scale=False, linesep=os.linesep,
```

File: plotille/_graphs.py [hunk 2]
```diff
@@ -31,6 +31,52 @@
...
+         lc=None, bg=None, color_mode='names'):
+    """
+    Create histogram for pre-aggregated data for existing bins and counts for each bin.
```
Comment (tammoippen) [inline plotille/_graphs.py:37]:
```suggestion
    Create histogram for aggregated data.
```

File: plotille/_graphs.py [hunk 3]
```diff
@@ -31,6 +31,52 @@
...
+
+    Parameters:
+        counts: List[int]         Count for each bucket / bn.
+        bins: List[float]    Bins for the provided counts. In case there are X counts, there should
+                             be X + 1 bins.
```
Comment (tammoippen) [inline plotille/_graphs.py:42]:
```suggestion
        counts: List[int]    Counts for each bucket.
        bins: List[float]    Limits for the bins for the provided counts: limits for 
                             bin `i` are `[bins[i], bins[i+1])`. 
                             Hence, `len(bins) == len(counts) + 1`.
```

File: plotille/_graphs.py [hunk 4]
```diff
@@ -31,6 +31,52 @@
...
+            ipf.fmt(b[i], delta=delta, chars=8, left=True),
+            ipf.fmt(b[i + 1], delta=delta, chars=8, left=False),
+            color('⣿' * (hight // 8) + lasts[hight % 8], fg=lc, bg=bg, mode=color_mode)
+            + color('\u2800' * (width - (hight // 8) + int(hight % 8 == 0)), bg=bg, mode=color_mode),
+            h[i])]
+    canvas += ['‾' * (2 * 8 + 2 + 3 + width + 12)]
+    return linesep.join(canvas)
+
 
 def hist(X, bins=40, width=80, log_scale=False, linesep=os.linesep,
```
Comment (tammoippen) [inline plotille/_graphs.py:81]: Implement the `hist` function using the new `hist_aggregated` function.

**Review by tammoippen — CHANGES_REQUESTED** [review]
Hi @Kami,
thank you for the interesting changes! I just have some remarks.

When you add a simple test and maybe an example where you get the aggregated data from an API or library, that would be great!

WRT documentation, i think you can copy most of the [hist doc](https://github.com/tammoippen/plotille#hist) and just add the changes.

**Comment by Kami** [comment]
@tammoippen Thanks for the review.

Do you think it would also make sense to implement the same changes for ``histogram()`` function?

I quickly looked at the implementation and that would require more re-factoring to get it to work (``Figure.histogram()``, ``Histogram`` class,, etc.).

**Comment by tammoippen** [comment]
> @tammoippen Thanks for the review.
> 
> Do you think it would also make sense to implement the same changes for `histogram()` function?
> 
> I quickly looked at the implementation and that would require more re-factoring to get it to work (`Figure.histogram()`, `Histogram` class, etc.).

I think for this PR the change to `hist` is enough. Afterwards, we can see, if we introduce aggregated variants to Figure.

**Comment by Kami** [comment]
I believe I've addressed all the feedback:

* Implemented `hist()` using `hist_aggregated()`
* Added a test case
* Added info on tje new function to the readme

Please let me know if I missing something.

**Review by tammoippen — APPROVED** [review]

**Comment by tammoippen** [comment]
Thank you for the PR. I will sort out the CI errors, they are not linked with your changes. 🎉

**Comment by Kami** [comment]
@tammoippen Thanks.

I also tried to get tests to work under Python 3.11, but it turns out it's hard to to because the project still supports Python 2.7 and there are some backward incompatible changes in newer versions of pytest which are needed for newer Python versions.

Having said that, I see you dropped support for 2.7 (#52), so I will revisit that change and see if I can make it to work.

**Pull request merged**